Rewrite UTF-8 validation in shift-based DFA for 70%~135% performance increase on non-ASCII strings

oxalica opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/136693)*
<!-- homu-ignore:end -->

Take 2 of rust-lang/rust#107760 (cc @thomcc)

### Background

About the technique: https://gist.github.com/pervognsen/218ea17743e1442e59bb60d29b1aa725

As stated in rust-lang/rust#107760,
> For prior art: shift-DFAs are now used for UTF-8 validation in [PostgreSQL](https://github.com/postgres/postgres/blob/aa6954104644334c53838f181053b9f7aa13f58c/src/common/wchar.c#L1753), and seems to be in progress or under consideration for use in https://github.com/JuliaLang/julia/pull/47880 and perhaps https://github.com/golang/go/issues/47120. Of these, PG's impl is the most similar to this one, at least at a high level[1](https://github.com/rust-lang/rust/pull/107760#user-content-fn-pg-a9b64c301fa792150e615ee4b94efb93).

### Rationales

1. Performance: This algorithm gives plenty of performance increase on validating strings with many non-ASCII codepoints, which is the normal case for almost every non-English content.

2. Generality: It does not use SIMD instructions and does not rely on the branch predictor to get a good performance, thus is good as a general, default, architecture-agnostic implementation. There is still a bypass for ASCII-only strings to benefit from auto-vectorization, if the target supports.

### Implementation details

I use the ordinary UTF-8 language definition from [RFC3692](https://datatracker.ietf.org/doc/html/rfc3629#section-4) and directly translate it into a 9-state DFA. ~~So the compressed state is 64-bit, resulting in a table of `[u64; 256]`, or 2KiB rodata.~~ Now the table is compressed into 1KiB `[u32; 256]`, see `library/core/src/str/solve_dfa.py` for details.

The main algorithm consists of following parts:
1. Main loop: taking a chunk of `MAIN_CHUNK_SIZE = 16` bytes on each iteration, execute the DFA on the chunk, and check if the state is in ERROR once per chunk.
2. ASCII bypass: in each chunk iteration, if the current state is ACCEPT, we know we are not in the middle of an encoded sequence, so we can skip a large block of trivial ASCIIs and stop at the first chunk containing any non-ASCII bytes. I choose `ASCII_CHUNK_SIZE = 16` to align with the current implementation: taking 16 bytes each to check non-ASCIIs, to encourage LLVM auto-vectorize it.
3. Trailing chunk and error reporting: execute the DFA step by step, stop on error as soon as possible, and calculate the error/valid location. To be simple, if any error are encountered in the main loop, it will discard the errornous chunk and `break` into this path to find the precise error location. That is, the erronous chunk, if exists, will be traversed twice, in exchange for a tighter and more efficient hot loop.

There are also some small tricks being used:
1. Since we have i686-linux in Tier 1 support, and its 64-bit shift (SHRD) is quite slow in our latency-intensive hot loop, I arrange the state storage so that the state transition can be done in 32-bit shift and conditional move. It shows a 200%+ speed up comparing to 64-bit-shift version.
2. We still need to get UTF-8 encoded length from the first byte in `utf8_char_width`. ~~I merge the previous lookup table into the unused high bits of the DFA transition table. So we don't need two tables. It did introduce an extra 32-bit shift. I believe it's almost free but have not benchmarked yet.~~ I kept `UTF8_CHAR_WIDTH` table unchanged.

### Benchmarks

I made an [out-of-tree implementation repository](https://github.com/oxalica/shift-dfa-utf8) for easier testing and benching. It also tested various `MAIN_CHUNK_SIZE` (m) and `ASCII_CHUNK_SIZE` (a) configurations. Bench data are taken from the first 4KiB (from the first paragraph, plain text not HTML, cut at char boundary) of Wikipedia [William Shakespeare in en](https://en.wikipedia.org/wiki/William_Shakespeare), [es](https://es.wikipedia.org/wiki/William_Shakespeare) and [zh](https://zh.wikipedia.org/wiki/%E5%A8%81%E5%BB%89%C2%B7%E8%8E%8E%E5%A3%AB%E6%AF%94%E4%BA%9A) language.

In short: with m=16, a=16, shift-DFA performance gives -45% on en, +69% on es, +135% on zh; with m=8, a=32, it gives +5% on en, +22% on es, +136% on zh. It's quite expected, as the larger ASCII bypass chunk is, it performs better on ASCII, but worse on mixed content like "es" because of the taken branch is flipping around.
To me, the difference in "en" is minimal in absolute time because the throughput is already high enough, comparing to not-as-fast "es". So I currently picking m=16, a=16 to lean towards "es" in the PR.

<details>
<summary>x86_64-linux results</summary>

On: Ryzen 7 5700G @3.775GHz (disabled turbo, with cpuset, with [stack layout randomizer](https://github.com/bheisler/criterion.rs/pull/744)):

Note: zh input consists of solely non-ASCII codepoints and runs fully on DFA path, it's the same performance as purely emoji inputs.

| Algorithm         | Input language | Throughput / (GiB/s)  |
|-------------------|----------------|-----------------------|
| std               | en             | 48.791 +-0.128        |
| shift-dfa-m16-a16 | en             | 26.815 +-0.006        | 
| shift-dfa-m8-a32  | en             | 51.314 +-0.018        |
| std               | es             |  5.925 +-0.069        |
| shift-dfa-m16-a16 | es             | 10.033 +-0.007        |
| shift-dfa-m8-a32  | es             |  7.254 +-0.035        |
| std               | zh             |  1.450 +-0.008        |
| shift-dfa-m16-a16 | zh             |  3.414 +-0.002        |
| shift-dfa-m8-a32  | zh             |  3.421 +-0.002        |


Before (486b0d11bd7d08f7a8f02f933b6447933f93009a):
```
    test string::from_utf8_lossy_100_ascii                   ... bench:          40.96 ns/iter (+/- 6.47)
    test string::from_utf8_lossy_100_invalid                 ... bench:       1,738.28 ns/iter (+/- 10.53)
    test string::from_utf8_lossy_100_multibyte               ... bench:          61.22 ns/iter (+/- 0.38)
    test string::from_utf8_lossy_invalid                     ... bench:         125.26 ns/iter (+/- 1.85)

    test str::str_validate_emoji                             ... bench:       3,279.32 ns/iter (+/- 54.01)
```
After with m16 a16 (ceb82dd971b7aef47493298255bab732bdc67b5e):
```
    test string::from_utf8_lossy_100_ascii                   ... bench:          23.92 ns/iter (+/- 0.10)
    test string::from_utf8_lossy_100_invalid                 ... bench:       1,962.22 ns/iter (+/- 42.19)
    test string::from_utf8_lossy_100_multibyte               ... bench:          41.19 ns/iter (+/- 0.75)
    test string::from_utf8_lossy_invalid                     ... bench:         155.69 ns/iter (+/- 3.46)

    test str::str_validate_emoji                             ... bench:       1,909.92 ns/iter (+/- 94.34)
```
After with m8 a32 (ceb82dd971b7aef47493298255bab732bdc67b5e):
```
    test string::from_utf8_lossy_100_ascii                   ... bench:          20.14 ns/iter (+/- 0.03)
    test string::from_utf8_lossy_100_invalid                 ... bench:       2,017.46 ns/iter (+/- 9.39)
    test string::from_utf8_lossy_100_multibyte               ... bench:          54.19 ns/iter (+/- 0.07)
    test string::from_utf8_lossy_invalid                     ... bench:         152.43 ns/iter (+/- 2.46)

    test str::str_validate_emoji                             ... bench:       2,715.99 ns/iter (+/- 164.68)
```
</details>


### Unresolved

- [x] Benchmark on aarch64-darwin, another tier 1 target.
  See [this comment](https://github.com/rust-lang/rust/pull/136693#issuecomment-2646355751).

- [ ] Decide the chunk size parameters. I'm currently picking m=16, a=16.

- [x] Should we also replace the implementation of [lossy conversion](https://github.com/oxalica/rust/blob/c0639b8cad126d886ddd88964f729dd33fb90e67/library/core/src/str/lossy.rs#L194) by calling the new validation function? It has a very similar code doing almost the same thing.

  It now also uses the new validation algorithm. Benchmarks of lossy conversions are included above.

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

oxalica committed:
Rewrite UTF-8 validation in shift-based DFA

This gives plenty of performance increase on validating strings with
many non-ASCII codepoints, which is the normal case for almost every
non-English content.

Shift-based DFA algorithm does not use SIMD instructions and does not
rely on the branch predictor to get a good performance, thus is good as
a general, default, architecture-agnostic implementation.
There is still a bypass for ASCII-only strings to benefits from
auto-vectorization, if the target supports.

We use z3 to find a state mapping that only need u32 transition table.
This shrinks the table size to 1KiB comparing a u64 states for less
cache pressure, and produces faster code on platforms that only support
32-bit shift. Though, it does not affect the throughput on 64-bit
platforms when the
table is already fully in cache.

hanna-kruppe commented on library/core/src/str/validations.rs:
```diff
@@ -112,172 +111,302 @@ where
+    let mut i = 0usize;
+
+    while i + MAIN_CHUNK_SIZE <= bytes.len() {
+        // Fast path: if the current state is ACCEPT, we can skip to the next non-ASCII chunk.
+        if st == ST_ACCEPT {
```
If I understand correctly, this tries to re-start the ASCII fast path after every `MAIN_CHUNK_SIZE` number of steps via the DFA, as long as the byte visited was the last byte of a UTF-8 encoded `char`. This can happen even if the entire input is 0% ASCII. In contrast, the old code only tries the ASCII fast path if it sees an ASCII byte, although it'll check after every single char. Presumably the strategy in this PR works pretty well in your benchmarks but I wonder if there's pathological inputs where it's... not necessarily worse than the old code, but at least not as good as it could be.

hanna-kruppe replied:
Yeah, on 0% ASCII inputs with frequent false starts in the ASCII fast path, I'd expect the exit on first chunk to be well predicted. Then the cost is only that more instructions have to be executed but aren't in the critical path as long as branch prediction is correct. However, it still isn't great:

1. Mixed ASCII/non-ASCII inputs (common in many real world scenarios) may make the ASCII-fast-path-exit less predictable, or at least require more precious branch predictor state to handle well.
2. The number of extra instructions executed (and brought into icache!) may be quite significant, especially on targets without suitable SIMD instructions, or if the vectorizer goes overboard with the setup code and unrolling. This probably won't show up in throughput oriented microbenchmarks (since the main DFA loop is ca. 1 cpb, there's enough spare issue slots to smooth it over) but it could hurt in more realistic scenarios when this cold is relatively cold and competing with other code.

Checking whether the first byte is ASCII seems like a very cheap way to not have to worry about it as much.

oxalica replied:
> 1. Mixed ASCII/non-ASCII inputs (common in many real world scenarios) may make the ASCII-fast-path-exit less predictable, or at least require more precious branch predictor state to handle well.

In this case, the branch flipping cannot be prevented by a 1byte inspection. But in the contrary, skipping ASCII chunks from a mixed input can actually improve performance, as long as the chunk size is not too big to give too many false positives. The "es" test covers this case.

> This probably won't show up in throughput oriented microbenchmarks (since the main DFA loop is ca. 1 cpb, there's enough spare issue slots to smooth it over) but it could hurt in more realistic scenarios when this cold is relatively cold and competing with other code.
> Checking whether the first byte is ASCII seems like a very cheap way to not have to worry about it as much.

Okay, I added the check. As it does not harm performance either. But I'd like to see some code illustrating this case for benchmarks.

hanna-kruppe replied:
> In this case, the branch flipping cannot be prevented by a 1byte inspection. But in the contrary, skipping ASCII chunks from a mixed input can actually improve performance, as long as the chunk size is not too big to give too many false positives. The "es" test covers this case.

Yes, if it's mixed then it'll drop in and out of the ASCII loop and that's probably a good thing if the ASCII runs are long enough. But during the *non-ASCII runs*, it seems preferable to have the not-perfectly-predictable branch be an up-front byte load + compare vs. having it be the exit condition of a chunky SIMD loop.



> But I'd like to see some code illustrating this case for benchmarks.

Unfortunately the only best way to put this sort of issue in the spotlight is to look at entire applications with an icache working set measured in megabytes (e.g, [the AsmDB paper](https://dl.acm.org/doi/pdf/10.1145/3307650.3322234) has an instructive story about memcmp in §4.4). If you want to try a microbenchmark, @the8472's example of parsing some binary format with utf8 validation as subroutine sounds like a good start, but I don't know if that will suffice to get you into the frontend-heavy regime.

hanna-kruppe commented on library/core/src/str/validations.rs:
```diff
@@ -112,172 +111,302 @@ where
+    TRANS_TABLE[byte as usize].wrapping_shr(st as _) as _
+}
+
+#[cfg(any(target_pointer_width = "16", target_pointer_width = "32"))]
```
Aside: there are a few targets that have 32 bit pointers but are capable of efficient 64 bit shifts. The most important once is wasm32, which has native 64 bit integer instructions and is often (but not always) executed on 64 bit capable hosts.

oxalica replied:
Fixed.

hanna-kruppe reviewed (commented):
Shift-based DFAs are a really cool trick, great to see that precise error reporting and ASCII fast paths can be added on top of them. I don't want to steal an actual reviewer's show and I've only given this a quick cursory read, but there's one thing I was wondering about and one small suggestion.

the8472 commented:
> or 2KiB rodata.

I suspect that this is going to eat into cache capacity, especially in code that interleaves string validation with other things, e.g. CBOR decoding. Have you benchmarked this for short inputs and with mixed workloads?

In coretests we have [TINY/SMALL/MEDIUM/LARGE/HUGE](https://github.com/rust-lang/rust/blob/master/library/coretests/benches/str/corpora.rs) for that reason

thaliaarchi commented:
> Should we also replace the implementation of [lossy conversion](https://github.com/rust-lang/rust/blob/c0639b8cad126d886ddd88964f729dd33fb90e67/library/core/src/str/lossy.rs#L194) by calling the new validation function?
> It has a very similar code doing almost the same thing.

FYI, in https://github.com/rust-lang/rust/pull/136677, I fixed the handling of invalid UTF-8 in `Display` for `OsStr`/`Path`, so there's a new user of lossy conversion. It needs to count the number of (potentially invalid) characters in a byte string and truncate it to some character width. I [extracted](https://github.com/rust-lang/rust/pull/136677/commits/d53ca9111f3140e05453d6fe59e49d76e8d20b70) a single iteration of the loop in `Utf8Chunks::next` to reuse it there. If `Utf8Chunks` is more amenable to using your optimization, `OsStr` might need a separate API.

hanna-kruppe commented:
> I suspect that this is going to eat into cache capacity, especially in code that interleaves string validation with other things, e.g. CBOR decoding.

This is a great point and should be benchmarked. Note that it's mitigated on many common cases by the accesses to the table depending only on the input bytes, so e.g. strings that all ASCII touch at most half the table. But that's still a lot of cache lines.

cuviper commented:
> I don't want to steal an actual reviewer's show

No worries on my part as the current assignee -- in fact, if another reviewer has more context here, I'd be happy to yield this one. Maybe @the8472 wants to take it?

oxalica commented:
@the8472 
> > or 2KiB rodata.
> 
> I suspect that this is going to eat into cache capacity, especially in code that interleaves string validation with other things, e.g. CBOR decoding. Have you benchmarked this for short inputs and with mixed workloads?

Could you please explain how to test "short inputs and with mixed workloads"? Should I bench the performance of a CBOR library like [ciborium](https://github.com/enarx/ciborium) using the new std on this branch? I'll give it a check later.

> In coretests we have [TINY/SMALL/MEDIUM/LARGE/HUGE](https://github.com/rust-lang/rust/blob/master/library/coretests/benches/str/corpora.rs) for that reason

It seems they are only used in little places like `str::char_count` and `str::debug`. `str::char()` is untouchedby this PR, because it uses the [UTF-8 decoder `next_code_point`](https://github.com/rust-lang/rust/blob/01e4f19cc8027925ffe0885a86388b700e46bfab/library/core/src/str/validations.rs#L36), not the validator. `str as Debug` is unrelated to validation either. This PR is only changing the implementation of `core::str::from_utf8` and its derivations. 

> Additionally you may want to benchmark 4-byte characters such as emoji or maybe some exotic scripts on the supplementary plane.

It should performs the same as the "zh" benchmark. Since the DFA is agnostic to the input, as long as it's purely non-ASCII to get into the DFA path, the performance should be the same. The correctness of DFA can be audited by checking state transitions table to match RFC3629.
Nevertheless, here's the benchmark (`./x.py bench --stage 0 library/coretests -- str::str_validate_emoji`):

```
Before:
str::str_validate_emoji                                         2742.95ns/iter   +/- 46.44
After:
str::str_validate_emoji                                         1529.48ns/iter    +/- 7.52
```

hanna-kruppe commented:
> No worries on my part as the current assignee -- in fact, if another reviewer has more context here, I'd be happy to yield this one.

I just meant that I, personally, couldn't take over the review even if I wanted to since I haven't had bors permissions in years 😄

hanna-kruppe commented:
Slightly silly idea, free to a good home: mapping the input bytes into equivalence classes before feeding them to the DFA, as done in [Flexible and Economical UTF-8 Decoder](https://bjoern.hoehrmann.de/utf-8/decoder/dfa/), would reduce the rodata size to 256B + 8B * (# classes). The critical path for the DFA transition is unaffected, so this would still run pretty fast. On Skylake-ish hardware, idealized peak throughput goes from 1 cycle per byte to 1.5cpb because the extra load becomes the bottleneck (that's for a loop that does nothing but compute the final state, the version in this PR doesn't seem to hit 1cpb anyway). On CPUs that can sustain more than two loads per cycle, it may not even be any slower for long inputs.

the8472 commented:
> It seems they are only used in little places like str::char_count and str::debug. str::char() is untouchedby this PR

Sorry, I didn't mean that this was specifically for utf8 validation, I was just gesturing at more cases that can be useful for string benchmarking in general to cover common scenarios.
Maybe you can look at earlier PRs that touched validation, sometimes other people have implemented their own external benchmark suites if I recall correctly.

> Could you please explain how to test "short inputs and with mixed workloads"? Should I bench the performance of a CBOR library like [ciborium](https://github.com/enarx/ciborium) using the new std on this branch? I'll give it a check later.

That's the tricky part. As hanna-kruppe says, such effects are most visible in complex codebases and are difficult to test with micro-benchmarks.

I suggested CBOR because it should contain short strings (e.g. dictionary keys) mixed with other non-string parsing and some implementations might already have their own benchmarks that can be reused. So it could be a benchmark of intermediate complexity.
You can try using `perf stat` to see if an existing benchmark is already experiencing cache pressure which might be exacerbated by the lookup table. Though I guess modern CPU caches may have some strategy that can deal with streaming workloads without clobbering the parts, so maybe it won't be complex enough.

I'll kick off a perf run to see if it impacts rustc itself. But we have already eliminated a bunch of string validations in the compiler, so it likely isn't a good benchmark either.

@bors try @rust-timer queue

>  Nevertheless, here's the benchmark (./x.py bench --stage 0 library/coretests -- str::str_validate_emoji):

👍

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 33c076ea1ecc54a73c507af70f6f2b8566b17c42 with merge 71c014723ade15a2bc2280eede6c97f96aa383fe...
<!-- homu: {"type":"TryBuildStarted","head_sha":"33c076ea1ecc54a73c507af70f6f2b8566b17c42","merge_sha":"71c014723ade15a2bc2280eede6c97f96aa383fe"} -->

oxalica commented:
@hanna-kruppe
> Slightly silly idea, free to a good home: mapping the input bytes into equivalence classes before feeding them to the DFA, as done in [Flexible and Economical UTF-8 Decoder](https://bjoern.hoehrmann.de/utf-8/decoder/dfa/), would reduce the rodata size to 256B + 8B * (# classes). That's 256B + 96B with the classes from the linked implementation, but idk if that number can be lower or has to be larger in this context.
> The critical path for the DFA transition is not affected by the input remapping

Not really. As I mentioned in the 32-bit-shift part, the bottleneck of the DFA path is latency, since the DFA has data dependency from each state to its precedent state. The later state can only be calculated after the previous result is out, no matter how many ALUs you have. So,

> goes from 1 cycle per byte to 1.5cpb

... literally means +50% latency and -50% performance. I tested it by introducing an identity lookup table `[u8; 256]` before each `TRANS_TABLE` lookup, then `throughput-zh-aligned/shift-dfa-m16-a16` goes down from 3.401GiB/s to 2.424GiB/s, which is -29%. So I don't think it worth it.

But yes, the table must be fully in cache to achieve these numbers, mixed workload is a challenge. I also prefer smaller table size, but not yet find a way to shrink it without introducing massive performance drop.

CryZe commented on library/core/src/str/validations.rs:
```diff
@@ -112,172 +112,308 @@ where
+const USE_SHIFT32: bool = cfg!(all(
+    any(target_pointer_width = "16", target_pointer_width = "32"),
+    // WASM32 supports 64-bit shift.
+    not(target_arch = "wasm32"),
```
You maybe want to include x86_64 and aarch64 here, which have targets with 32-bit pointers.

oxalica replied:
> You maybe want to include x86_64 and aarch64 here, which have targets with 32-bit pointers.

How can you target x86_64 and aarch64 but forcing 32-bit pointers? I don't know it's possible.

hanna-kruppe replied:
By using an ABI with 32 bit pointers, such as [x32](https://en.wikipedia.org/wiki/X32_ABI). I don't think any of those are popular today (certainly not tier 1-2 targets in rustc) but they do exist.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13207672531/job/36878905776)
Build commit: 71c014723ade15a2bc2280eede6c97f96aa383fe (`71c014723ade15a2bc2280eede6c97f96aa383fe`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13207672531/job/36878905776"},"merge_sha":"71c014723ade15a2bc2280eede6c97f96aa383fe"} -->

rust-timer commented:
Queued 71c014723ade15a2bc2280eede6c97f96aa383fe with parent 64e06c0f5578829373743884b708d494136c3e8f, future [comparison URL](https://perf.rust-lang.org/compare.html?start=64e06c0f5578829373743884b708d494136c3e8f&end=71c014723ade15a2bc2280eede6c97f96aa383fe).
There are currently 2 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~3.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

oxalica commented:
@the8472 
> > Could you please explain how to test "short inputs and with mixed workloads"? Should I bench the performance of a CBOR library like [ciborium](https://github.com/enarx/ciborium) using the new std on this branch? I'll give it a check later.
> 
> That's the tricky part. As hanna-kruppe says, such effects are most visible in complex codebases and are difficult to test with micro-benchmarks.
> 
> I suggested CBOR because it should contain short strings (e.g. dictionary keys) mixed with other non-string parsing and some implementations might already have their own benchmarks that can be reused. So it could be a benchmark of intermediate complexity.

I wrote some CBOR deserialization [bench in my out-of-tree repo](https://github.com/oxalica/shift-dfa-utf8/blob/main/bench-de/benches/cbor.rs). By running it on this PR and master rustc sysroot, [the result](https://github.com/oxalica/shift-dfa-utf8/blob/21be3928febcdb9449a6cf32361d5a4e59ce0a23/bench-de/benchmark.txt) shows this PR gives an +0.3%~1.4% performance improvement. On small-string-heavy `twitter.json.dagcbor` the difference is negligible. I doubt that these numbers are mostly just noise and there is no significant change in performance.

joboet commented:
I think it's worth mentioning that in https://github.com/rust-lang/rust/pull/107760 @thomcc managed to find a way to pack the 9 states of the DFA into 32-bits, which would help address the cache size, binary size and portability concerns. Perhaps you could ask him if he'd provide you with the SMT-solver code he used to find the state values?

rust-timer commented:
Finished benchmarking commit (71c014723ade15a2bc2280eede6c97f96aa383fe): [comparison URL](https://perf.rust-lang.org/compare.html?start=64e06c0f5578829373743884b708d494136c3e8f&end=71c014723ade15a2bc2280eede6c97f96aa383fe&stat=instructions:u).

### Overall result: ❌ regressions - please read the text below


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=64e06c0f5578829373743884b708d494136c3e8f&end=71c014723ade15a2bc2280eede6c97f96aa383fe&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.3% | [3.3%, 3.3%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 3.3% | [3.3%, 3.3%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=64e06c0f5578829373743884b708d494136c3e8f&end=71c014723ade15a2bc2280eede6c97f96aa383fe&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=64e06c0f5578829373743884b708d494136c3e8f&end=71c014723ade15a2bc2280eede6c97f96aa383fe&stat=cycles:u)
<details>
<summary>Results (primary 3.3%, secondary 2.5%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.3% | [3.3%, 3.3%] | 1     |
| Regressions ❌ <br /> (secondary)  | 2.5% | [1.4%, 2.9%] | 9     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 3.3% | [3.3%, 3.3%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=64e06c0f5578829373743884b708d494136c3e8f&end=71c014723ade15a2bc2280eede6c97f96aa383fe&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.2%, secondary 0.5%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2% | [0.1%, 0.5%] | 14    |
| Regressions ❌ <br /> (secondary)  | 0.5% | [0.3%, 0.6%] | 38    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.2% | [0.1%, 0.5%] | 14    |
</details>

**Bootstrap**: 779.143s -> 778.734s (-0.05%)
**Artifact size**: 329.04 MiB -> 329.08 MiB (0.01%)

oxalica committed:
Align transition table and fit it in a single page

1. To reduce the cache footprint.
2. To avoid additional cost when access across pages.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13221556809/job/36907231756) [(plain)](https://github.com/rust-lang/rust/commit/9cc648bad370cec093e6326820eff8d1c8cd644b/checks/36907231756/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04
ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
  make \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 2.939 Building wheels for collected packages: reuse
#12 2.941   Building wheel for reuse (pyproject.toml): started
#12 3.161   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.162   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=be6760d5849de4a58bbe52b85ca57a55f2b32b518b17029a5ad2e530db0d4303
#12 3.162   Stored in directory: /tmp/pip-ephem-wheel-cache-f1rya2_y/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.165 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.582 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.582 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.142 Collecting virtualenv
#12 4.142 Collecting virtualenv
#12 4.189   Downloading virtualenv-20.29.1-py3-none-any.whl (4.3 MB)
#12 4.266      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 57.5 MB/s eta 0:00:00
#12 4.309 Collecting distlib<1,>=0.3.7
#12 4.314   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.322      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 81.7 MB/s eta 0:00:00
#12 4.360 Collecting filelock<4,>=3.12.2
#12 4.365   Downloading filelock-3.17.0-py3-none-any.whl (16 kB)
#12 4.400 Collecting platformdirs<5,>=3.9.1
#12 4.405   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#12 4.488 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.675 Successfully installed distlib-0.3.9 filelock-3.17.0 platformdirs-4.3.6 virtualenv-20.29.1
#12 DONE 4.8s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      128960 kB
DirectMap2M:     6162432 kB
DirectMap1G:    12582912 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2025-01-08/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
fmt check
fmt: checked 5834 files
tidy check
tidy: Skipping binary file check, read-only filesystem
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.3.1)
  Downloading pip-25.0-py3-none-any.whl.metadata (3.7 kB)
Downloading pip-25.0-py3-none-any.whl (1.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 106.4 MB/s eta 0:00:00
Installing collected packages: pip
---
Successfully installed pip-25.0
linting python files
All checks passed!
checking python file formatting
Would reformat: /checkout/library/core/src/str/solve_dfa.py

python formatting does not match! Printing diff:
1 file would be reformatted, 28 files already formatted
1 file would be reformatted, 28 files already formatted
--- /checkout/library/core/src/str/solve_dfa.py
+++ /checkout/library/core/src/str/solve_dfa.py
 ]
 
 o = Optimize()
 o = Optimize()
-offsets = [BitVec(f'o{i}', 32) for i in range(STATE_CNT)]
-trans_table = [BitVec(f't{i}', 32) for i in range(len(TRANSITIONS))]
+offsets = [BitVec(f"o{i}", 32) for i in range(STATE_CNT)]
+trans_table = [BitVec(f"t{i}", 32) for i in range(len(TRANSITIONS))]
 # Add some guiding constraints to make solving faster.
 # Add some guiding constraints to make solving faster.
 o.add(offsets[0] == 0)
 o.add(trans_table[-1] == 0)
 
 for i in range(len(offsets)):
-    o.add(offsets[i] < 32 - 5) # Do not over-shift. It's not necessary but makes solving faster.
+    o.add(
+        offsets[i] < 32 - 5
+    )  # Do not over-shift. It's not necessary but makes solving faster.
     for j in range(i):
         o.add(offsets[i] != offsets[j])
 for trans, (targets, _) in zip(trans_table, TRANSITIONS):
@@ -54,10 +56,10 @@
 goal = Concat(*offsets, *trans_table)
 o.minimize(goal)
 print(o.check())
-print('Offset[]= ', [o.model()[i].as_long() for i in offsets])
-print('Transitions:')
+print("Offset[]= ", [o.model()[i].as_long() for i in offsets])
+print("Transitions:")
 for (_, label), v in zip(TRANSITIONS, [o.model()[i].as_long() for i in trans_table]):
-    print(f'{label:14} => {v:#10x}, // {v:032b}')
+    print(f"{label:14} => {v:#10x}, // {v:032b}")
 # Output should be deterministic:
 # sat


tidy error: checks with external tool 'ruff' failed
rerun tidy with `--extra-checks=py:fmt --bless` to reformat Python code
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Sun Feb  9 02:40:59 UTC 2025
  network time: Sun, 09 Feb 2025 02:40:59 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

oxalica commented:
@joboet

> I think it's worth mentioning that in #107760 @thomcc managed to find a way to pack the 9 states of the DFA into 32-bits, which would help address the cache size, binary size and portability concerns. Perhaps you could ask him if he'd provide you with the SMT-solver code he used to find the state values?

Thanks for pointing out this and I updated the patch to use a u32 table now. I checked the script mentioned in postgres comments and rewrote it for our use case:
1. Use the state assignment mentioned in this PR previously, which is a sequential assignment on the language definition, not using their newly-invented state names.
2. Optimize on the solution numbers to get the minimal solution, so it's deterministic and reproducible. It should also be stable across SAT solver used and/or its version.

I put the generating python script in tree and appended the output solution to it for reference. I'm not sure if there is a better place to put it, but I found there is [`library/core/src/unicode/printable.py`](https://github.com/oxalica/rust/blob/01e4f19cc8027925ffe0885a86388b700e46bfab/library/core/src/unicode/printable.py). So I guess putting it next to the relevant Rust code is fine?

Performance wise, u32 transitions make the algorithm on i686 runs almost as fast as on x86_64. But it does not affect much on x86_64 alone. It should reduce cache pressure but that may not be easily benchmark-able.

@the8472 
I'm not sure how to interpret the numbers in the [perf report](https://github.com/rust-lang/rust/pull/136693#issuecomment-2644415987). Is it considered a severe regression? I think that the slower compilation time is mainly affected by slower ASCII checking, as rustc itself does only ASCII processing and most Rust source codes are in ASCII. If we want to optimize for this, I'll change the chunking parameters to `MAIN_CHUNK_SIZE=8, ASCII_CHUNK_SIZE=32`.

All benchmark results are updated in my repo: [x86_64](https://github.com/oxalica/shift-dfa-utf8/blob/7c3c10aa081362fb7f21334390d92b1dc4fdae42/result-x86_64.txt), [i686](https://github.com/oxalica/shift-dfa-utf8/blob/7c3c10aa081362fb7f21334390d92b1dc4fdae42/result-i686.txt), [CBOR deserialization](https://github.com/oxalica/shift-dfa-utf8/blob/7c3c10aa081362fb7f21334390d92b1dc4fdae42/bench-de/result-x86_64.txt) compiled by rustc from this PR.

the8472 commented:
> I'm not sure how to interpret the numbers in the [perf report](https://github.com/rust-lang/rust/pull/136693#issuecomment-2644415987). Is it considered a severe regression?

Syn has been noisy, so that one can be ignored which means instructions look neutral.

But cycles look like they regressed across a few places. tuple-stress is spending more cycles across several variations of the benchmark... and yet instruction counts are unchanged. Weird. But it's a stress test so as long as it's not a huge impact it's probably not relevant.

Runtime benchmarks also aren't showing anything that would obviously point at validation. Brotli should be munging bytes, not strings.

If you want to doublecheck you could try running those benchmarks under perf diff locally and see if anything string-related shows up, but I doubt that it will.

Anyway, since the impl changed, let's see rerun perf.

@bors try @rust-timer queue

> All benchmark results are updated in my repo: [x86_64](https://github.com/oxalica/shift-dfa-utf8/blob/7c3c10aa081362fb7f21334390d92b1dc4fdae42/result-x86_64.txt), [i686](https://github.com/oxalica/shift-dfa-utf8/blob/7c3c10aa081362fb7f21334390d92b1dc4fdae42/result-i686.txt), [CBOR deserialization](https://github.com/oxalica/shift-dfa-utf8/blob/7c3c10aa081362fb7f21334390d92b1dc4fdae42/bench-de/result-x86_64.txt) compiled by rustc from this PR.

Have you run them several times to check if there's variance between runs? ASLR, cpu clock boosting, differences between CPU cores etc. can lead to stable within-run but varying cross-run results.

Assuming they are stable, it looks like

- throughput is generally up in microbenchmarks
- but throughput is a bit worse for `cbor-de-citm_catalog.json`
- on 32bit latency is worse than std for short strings
- on 64bit latency looks mixed... 
- `cbor-de-trivial_helloworld` (another short string case) ends up worse too

That there are some latency improvements at all on x86-64 is tantalizing. But they appear to evaporate once it's embedded in additional code (for the single context, CBOR, that's being tested here). So maybe the DFA algorithm should only be used for strings above a certain length?

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 486b0d11bd7d08f7a8f02f933b6447933f93009a with merge 0de538c8d4bb71779021767a5bd1e3a72da0b09b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"486b0d11bd7d08f7a8f02f933b6447933f93009a","merge_sha":"0de538c8d4bb71779021767a5bd1e3a72da0b09b"} -->

kornelski commented:
[Benchmark on Apple M3 Max](https://github.com/user-attachments/files/18725009/result-apple-m3max.txt)

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13226558196/job/36920071986)
Build commit: 0de538c8d4bb71779021767a5bd1e3a72da0b09b (`0de538c8d4bb71779021767a5bd1e3a72da0b09b`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13226558196/job/36920071986"},"merge_sha":"0de538c8d4bb71779021767a5bd1e3a72da0b09b"} -->

rust-timer commented:
Queued 0de538c8d4bb71779021767a5bd1e3a72da0b09b with parent 1ff21350fdc967393243227c88dddb5def8717b2, future [comparison URL](https://perf.rust-lang.org/compare.html?start=1ff21350fdc967393243227c88dddb5def8717b2&end=0de538c8d4bb71779021767a5bd1e3a72da0b09b).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.8 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

oxalica commented:
@the8472 
> Syn has been noisy, so that one can be ignored which means instructions look neutral.
> 
> But cycles look like they regressed across a few places. tuple-stress is spending more cycles across several variations of the benchmark... and yet instruction counts are unchanged. Weird. But it's a stress test so as long as it's not a huge impact it's probably not relevant.
> 
> Runtime benchmarks also aren't showing anything that would obviously point at validation. Brotli should be munging bytes, not strings.
> 
> If you want to doublecheck you could try running those benchmarks under perf diff locally and see if anything string-related shows up, but I doubt that it will.
> 
> Anyway, since the impl changed, let's see rerun perf.

Thanks for the explanation. I'll play it around locally.

> Have you run them several times to check if there's variance between runs? ASLR, cpu clock boosting, differences between CPU cores etc. can lead to stable within-run but varying cross-run results.

All my benchmarks are produced under my [`cbench` tool](https://github.com/oxalica/cbench) which does control most points you mentioned: disalbe ASLR, disable CPU boosting, lock cpufreq, cpuset on a single core and forbid other programs to run on it, set irq affinity to avoid that core, and disable its hyper-threading sibling core. The result is pretty stable in a single run (as seen by the small deviation), but it can still fluctuate a little (mostly <1%) across different runs, which may blame memory layout change as mentioned in [this criterion issue](https://github.com/bheisler/criterion.rs/issues/334).


> * throughput is generally up in microbenchmarks
> 
>     * but throughput is a bit worse for `cbor-de-citm_catalog.json`
> 
>     * on 32bit latency is worse than std for short strings
> 
>     * on 64bit latency looks mixed...
> 
>     * `cbor-de-trivial_helloworld` (another short string case) ends up worse too

I'm not expecting to see a performance increase in these mixed workload. But I'll look into these, especially the helloworld one.


@kornelski 

> Benchmark on Apple M3 Max

Thanks for the data. The result seems quite similar:

1. The ASCII case results in smaller difference between different chunking parameters. But the relative trend is quite similar.
2. Aligned and unaligned performances are also similar enough to be ignored, as in x86_64.
3. The latency result is a lot better than x86_64, so the absolute difference is smaller (1-2ns diff). But this patch behaves generally worse than std now, probably because the branch is better predicted in std?

I came up an idea to move the partial chunk processing from the tail to the head, to bless it to help in small string cases. I'll test it locally first.

rust-timer commented:
Finished benchmarking commit (0de538c8d4bb71779021767a5bd1e3a72da0b09b): [comparison URL](https://perf.rust-lang.org/compare.html?start=1ff21350fdc967393243227c88dddb5def8717b2&end=0de538c8d4bb71779021767a5bd1e3a72da0b09b&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=1ff21350fdc967393243227c88dddb5def8717b2&end=0de538c8d4bb71779021767a5bd1e3a72da0b09b&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=1ff21350fdc967393243227c88dddb5def8717b2&end=0de538c8d4bb71779021767a5bd1e3a72da0b09b&stat=max-rss)
<details>
<summary>Results (primary 0.8%, secondary -1.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.8%  | [0.8%, 0.8%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.9% | [-1.9%, -1.9%] | 1     |
| All ❌✅ (primary)                 | 0.8%  | [0.8%, 0.8%]   | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=1ff21350fdc967393243227c88dddb5def8717b2&end=0de538c8d4bb71779021767a5bd1e3a72da0b09b&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=1ff21350fdc967393243227c88dddb5def8717b2&end=0de538c8d4bb71779021767a5bd1e3a72da0b09b&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.2%, secondary 0.6%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2% | [0.1%, 0.5%] | 12    |
| Regressions ❌ <br /> (secondary)  | 0.6% | [0.3%, 0.7%] | 38    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.2% | [0.1%, 0.5%] | 12    |
</details>

**Bootstrap**: 781.536s -> 778.817s (-0.35%)
**Artifact size**: 329.13 MiB -> 329.11 MiB (-0.01%)

the8472 commented:
This is looking quite neutral now. Combined with the throughput improvements in microbenchmarks that should be good enough for inclusion as far as perf aspects go.

oxalica committed:
Reuse core::str::from_utf8 in lossy UTF-8 parsing

oxalica committed:
Process partial chunks at beginning and remove `unlikely` hints

Hope to have a better latency on short strings and/or the immediate-fail
path.

oxalica committed:
Use UTF-8 error length enum to reduce register spill

When using `error_len: Option<u8>`, `Result<(), Utf8Error>` will be
returned on stack and produces suboptimal stack suffling operations. It
causes 50%-200% latency increase on the error path.

oxalica commented:
I updated lossy parsing/conversion (`core::str::Utf8Chunks::next`) to use the new validation algorithm. It took some extra work to perform relatively well on error latency, without suffering a 200% regression due to LLVM's bad register allocation (bug?) and a mild suboptimal codegen #136972.

On d2030aa5afdf483789ce4d09c1a6f82ea4e47287, It shows a +11%~49% speed up on valid to almost-valid path, and a -13% regresson on worst path (all bytes are invalid). The number seems acceptable to me.

joboet reviewed (commented):
Thank you for pushing this further! I have some style nits...

Also, I *think* (correct me if I'm wrong) you could get rid of most of the `unsafe` stuff by rewriting the main loop around an iterator over `as_rchunks()` and computing the position for `resolve_error_location` using the length of the remaining slice like so:
```rust
let (remainder, chunks) = bytes.as_rchunks();
// ... check the remainder ...

let mut chunks = chunks.iter();
while let Some(mut chunk) = chunks.next() {
    if st == ST_ACCEPT && chunk[0].is_ascii() {
        while chunk_is_ascii {
            chunk = chunks.next();
        }
    } else {
        // ... check chunk ...

        if error {
            let i = bytes.len() - chunks.as_slice().as_flattened().len() - CHUNK_SIZE;
            // handle error
        }
    }
}
```
This is completely equivalent to the current version, doesn't introduce bound checks and the only `unsafe` remaining would be in `run_with_error_handling`.

oxalica commented:
> Also, I _think_ (correct me if I'm wrong) you could get rid of most of the `unsafe` stuff by rewriting the main loop around an iterator over `as_rchunks()` and computing the position for `resolve_error_location` using the length of the remaining slice like so:

It only works if `MAIN_CHUNK_SIZE == ASCII_CHUNK_SIZE` I think? Currently they are both 16, but I think m8 a16 is another candidate with different pros&cons, and haven't really decide yet.

the8472 commented:
For chunking into arrays of varying sizes a `split_off_chunk_last` method would help, but currently that's not included in #136693

oxalica committed:
Prefer pure functions over mutable arguments

oxalica committed:
Fix comments, use `u8::is_ascii` and simplify

bors commented:
:umbrella: The latest upstream changes (presumably #138155) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

oxalica committed:
Add a benchmark of UTF-8 validation for ASCII

oxalica committed:
Only use non-chunked UTF8 validation on `optimize_for_size`

oxalica commented:
I pushed a change to use the simple loop without fast-path or loop-unrolling under `optimize_for_size`. And I checked the binary size of generated code under size optimization suggested by [min-sized-rust](https://github.com/johnthagen/min-sized-rust), with various profile settings and `-Z build-std=std,panic_abort -Z build-std-features=panic_immediate_abort`.

The `.so` under x86_64-linux is 4528B on nightly 2025-03-01, and 5928B after this PR (https://github.com/rust-lang/rust/pull/136693/commits/ff0055905d2a5e7a47d9581c9ba6eea3082119e4). So it's about +1.37KiB, most of which is the 1KiB `[u32; 256]` state transition table, as expected.

Test code:

```rust
// crate-type = ["cdylib"]
#![no_std]

#[no_mangle]
pub unsafe extern "C" fn validate_utf8(ptr: *const u8, len: usize) -> bool {
    let bytes = core::slice::from_raw_parts(ptr, len);
    core::str::from_utf8(bytes).is_ok()
}
```

joboet commented on library/core/src/str/validations.rs:
Shouldn't this be
```suggestion
        if st & STATE_MASK == ST_ACCEPT && unsafe { bytes.get_unchecked(i).is_ascii() } {
```

oxalica replied:
Thanks for the catch. Fixed.

oxalica committed:
Fix missed state masking

CrazyboyQCD commented:
This implementation seems not that effective as it should be
https://godbolt.org/z/b638vffz6
The Zig implementation comes from https://github.com/ziglang/zig/compare/e0d3885f0b98f999e5a17814b47a50845cd2caea..fb482e1e27b6da790c6fc501692bac24178250db
https://github.com/ziglang/zig/blob/fb482e1e27b6da790c6fc501692bac24178250db/lib/std/unicode.zig

oxalica commented:
> This implementation seems not as effective as it should be
> https://godbolt.org/z/zsqbseYhP
> The Zig implementation comes from https://github.com/ziglang/zig/compare/e0d3885f0b98f999e5a17814b47a50845cd2caea..fb482e1e27b6da790c6fc501692bac24178250db

That's not comparable because your Zig implementation returns a simple bool, but the Rust implementation does exact error location resolution and error length detection. For your inputs containing errors, more than 60% time is spent on error path: the unrolled loop fails and we need to re-parse the last chunk with one branch each iteration to get the exact error location and the "error length". Note that the error path is also why we cannot get chunk size too large to utilize larger SIMD width. (And your inputs is completely ASCII, which misses the point of this PR to optimize non-ASCII strings. For ASCII, the DFA path is completely useless except for the final chunk.)

That's a plenty of cost for a one-shot validation. But I don't think we can eliminate these because the behavior is documented by [`Utf8Error`](https://doc.rust-lang.org/1.87.0/std/str/struct.Utf8Error.html) and there are many users of the precise location (eg, [`<[u8]>::utf8_chunks`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.utf8_chunks)).

By changing to return `Result<(), ()>`, and swap the remainder chunk process to the end, it's 50% faster than old impl. I moved the remainder chunk to the start just for fail-fast purpose https://github.com/rust-lang/rust/pull/136693/commits/46a4782d33468c569d7d9001580c6fd13f098934 due to the chunk re-parse penalty on completely invalid inputs. If we do not resolve error locations, it's also unnecessary.
https://godbolt.org/z/WsPznxhj4
It's still slower than Zig though, possibly because of wider SIMD ops, etc. I don't think we can directly follow these optimizations within the limitation of our error API.

bors commented:
:umbrella: The latest upstream changes (presumably #143521) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

oxalica commented:
I'm closing this because: 
1. I'm unsure if this general direction would be acceptable, considering the complexity of transition table construction. And it is hard to reason or verify its correctness without an external tool (z3). 
2. I'm unhappy about the current implementation containing so many special cases and hacks: ASCII-bypass is okay but it does not blend well with the rest of the algorithm. Auto-vectorization and unrolling also need tricks for LLVM to produce nice code. Error recovery and precise error location extraction logic are lengthy and slow, giving a huge penalty on `Err` cases.
3. The performance after this change is not so impressive and is still on the same magnitude. It cannot leverage SIMD either, though SIMD algorithm like simdutf8 has its own issue as explained in #107760. But I would like to see a more parallel algorithm nowadays, comparing to strictly-serial DFA which cannot pass 1byte/cycle boundary[^1].

Feel free to pick it up if someone still has interests on this approach.

[^1]: Theoretically you can, by doing instruction-level parallelism, but no.
